### PR TITLE
Rewards: Lock rule as long as precondition rule is not unlocked

### PR DESCRIPTION
**Pull request by sparrowDom**
### Description:
- Create a template for august campaign.
- Add a social charing boilerplate rule code.
- Add a functionality that enables you to lock a rule by conditioning it with completion of another rule

### Checklist:

- [ ] Test your work and double-check to confirm that you didn't break anything
- [ ] [Wrap any new text/strings with `fbt`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#wrapping-text) for translation
- [ ] If there are any new/changed strings to translate, [`npm run translate`](https://github.com/OriginProtocol/origin/tree/master/dapps/marketplace/translation#integrating-translations)
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/docs)

### CODE REVIEW

File: infra/growth/src/util/arrayUtils.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+function concat(x, y) {
+  return x.concat(y)
+}
+
+function flatMap(f, xs) {
+  return xs.map(f).reduce(concat, [])
```
Comment (shahthepro) [inline infra/growth/src/util/arrayUtils.js]: Just a note. We can probably skip the `.map` here.
```
xs.reduce((a, d) => [...a, ...f(d)], [])
```
But your approach looks pretty cool.

File: infra/growth/src/resources/rules.js [hunk 2]
```diff
@@ -465,13 +504,33 @@ class BaseRule {
...
    */
-  async getStatus(ethAddress, events, currentUserLevel) {
+  async getStatus(ethAddress, events, currentUserLevel, rules) {
     // If the user hasn't reached the level the rule is in, status is Inactive.
     if (currentUserLevel < this.levelId) {
       return GrowthActionStatus.Inactive
     }
+    // If the rule has additional unlock conditions
+    else if (this.config.additionalLockConditions) {
+      this._validateLockConditions(rules, this.config.additionalLockConditions)
```
Comment (franckc) [inline infra/growth/src/resources/rules.js]: As opposed to calling _validateLockConditions here, could we call it in the constructor of the rule ? This way we wouldn't wait for getStatus to get called to find configuration issues.

File: infra/growth/src/resources/rules.js [hunk 3]
```diff
@@ -231,10 +232,18 @@ class CampaignRules {
...
     const data = []
     for (let i = 0; i < this.numLevels; i++) {
       data.push(
-        ...(await this.levels[i].export(adapter, ethAddress, events, level))
+        ...(await this.levels[i].export(
+          adapter,
+          ethAddress,
+          events,
+          level,
+          allRules
```
Comment (franckc) [inline infra/growth/src/resources/rules.js]: Do we really need to pass this additional parameters ? Both Level and BaseRule have a "crules" field that gives access to all the rules. So the rules should be already accessible without this extra parameter.

File: infra/growth/src/resources/rules.js [hunk 4]
```diff
@@ -868,4 +928,36 @@ class ReferralRule extends BaseRule {
   }
 }
 
+/**
+ * A rule that requires the purchase of a specific listing.
```
Comment (franckc) [inline infra/growth/src/resources/rules.js]: nit: fix comment

File: infra/growth/src/util/arrayUtils.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+}
+
+module.exports = { concat, flatMap }
```
Comment (franckc) [inline infra/growth/src/util/arrayUtils.js]: nit: no need to export concat.

File: infra/growth/src/util/arrayUtils.js (hunk 1 above)
Comment (sparrowDom) [reply infra/growth/src/util/arrayUtils.js]: Thanks! Skipping the concat function like suggested. 

About the `map` I kind of prefer keeping my version. While your works nicely and uses less code I think with `map` the code is a bit more readable.

File: infra/growth/src/util/arrayUtils.js (hunk 5 above)
Comment (sparrowDom) [reply infra/growth/src/util/arrayUtils.js]: tnx, incorporating @shahthepro's change the concat is removed

File: infra/growth/src/resources/rules.js (hunk 4 above)
Comment (sparrowDom) [reply infra/growth/src/resources/rules.js]: 👍

File: infra/growth/src/resources/rules.js (hunk 3 above)
Comment (sparrowDom) [reply infra/growth/src/resources/rules.js]: good point! Time to un-mess the method signatures

File: infra/growth/src/resources/rules.js (hunk 2 above)
Comment (sparrowDom) [reply infra/growth/src/resources/rules.js]: Yup good point thanks. Moved to the constructor

**Comment by sparrowDom** [comment]
Thanks @franckc  & @shahthepro for the CR. I haven't added any tests for the SocialSharing rule, since it is just boilerplate code ATM. 

There are tests for invalid locking configuration, and for checking that the additional lock conditions functionality is working. Both of those are at the end of `apollo-adapter-august.test.js`

All the CR stuff has been addressed so merging this for now.

**Pull request merged**